compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -315,7 +315,7 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             return ambig;
         }
 
-        let mut candidates = self.assemble_candidates_via_self_ty(goal);
+        let mut candidates = self.assemble_candidates_via_self_ty(goal, 0);
```

Comment:
this could be tracked on the eval ctxt

Reply:
tracking it in the `EvalCtxt` was also somewhat ugly so I didn't do that. I think that we may want to also track all visited self types as a hashset or sth here in the future and expect that to be nicer by passing the current data in.